src/chunk_index.c

```diff
@@ -102,35 +103,93 @@ chunk_index_choose_name(const char *tabname, const char *main_index_name, Oid na
 	return idxname;
 }
 
+static inline AttrNumber
+find_attno_by_name(TupleDesc tupdesc, Name attname)
+{
+	int			i;
+
+	for (i = 0; i < tupdesc->natts; i++)
+	{
+		FormData_pg_attribute *attr = tupdesc->attrs[i];
+
+		if (strncmp(NameStr(attr->attname), NameStr(*attname), NAMEDATALEN) == 0)
+			return attr->attnum;
+	}
+	return InvalidAttrNumber;
+}
+
 /*
- * Translate a hypertable's index attribute numbers to match a chunk.
- *
- * A hypertable's IndexInfo for one of its indexes references the attributes
- * (columns) in the hypertable by number. These numbers might not be the same
- * for the corresponding attribute in one of its chunks. To be able to use an
- * IndexInfo from a hypertable's index to create a corresponding index on a
- * chunk, we need to translate the attribute numbers to match the chunk.
+ * Adjust attribute numbers for expression index definitions.
  */
 static void
-chunk_translate_attnos(IndexInfo *ii, Relation idxrel, Relation chunkrel)
+chunk_adjust_expr_attnos(IndexInfo *ii, Relation htrel, Relation idxrel, Relation chunkrel)
+{
+	ListCell   *lc;
+
+	foreach(lc, ii->ii_Expressions)
+	{
+		/* Get a list of references to all Vars in the expression */
+		List	   *vars = pull_var_clause(lfirst(lc), 0);
+		ListCell   *lc_var;
+		int			i;
+
+		foreach(lc_var, vars)
+		{
+			Var		   *var = lfirst(lc_var);
+			bool		found = false;
+
+			/*
+			 * Find the hypertable attribute that matches the Var. Then, given
+			 * the attribute's name, find the chunk attribute that matches.
+			 * Finally, adjust the Var's attno to match the chunk's attno for
+			 * that attribute.
+			 */
+			for (i = 0; i < htrel->rd_att->natts; i++)
+			{
+				FormData_pg_attribute *htattr = htrel->rd_att->attrs[i];
+
+				if (htattr->attnum == var->varattno)
+				{
+					AttrNumber	attno = find_attno_by_name(chunkrel->rd_att, &htattr->attname);
+
+					if (attno != InvalidAttrNumber)
+					{
+						var->varattno = attno;
+						found = true;
+						break;
+					}
+				}
+			}
+
+			if (!found)
+				elog(ERROR, "Index attribute %s not found in chunk",
+					 NameStr(idxrel->rd_att->attrs[i]->attname));
+		}
+	}
+}
+
+/*
+ * Adjust column reference attribute numbers for regular indexes.
+ */
+static void
+chunk_adjust_colref_attnos(IndexInfo *ii, Relation idxrel, Relation chunkrel)
 {
 	int			i,
 				natts = 0;
 
-	Assert(ii->ii_NumIndexAttrs == idxrel->rd_att->natts);
-
 	for (i = 0; i < idxrel->rd_att->natts; i++)
 	{
+		FormData_pg_attribute *idxattr = idxrel->rd_att->attrs[i];
 		bool		found = false;
 		int			j;
 
 		for (j = 0; j < chunkrel->rd_att->natts; j++)
 		{
-			if (strncmp(NameStr(idxrel->rd_att->attrs[i]->attname),
-						NameStr(chunkrel->rd_att->attrs[j]->attname),
-						NAMEDATALEN) == 0)
+			AttrNumber	attno = find_attno_by_name(chunkrel->rd_att, &idxattr->attname);
+
+			if (attno != InvalidAttrNumber)
 			{
-				ii->ii_KeyAttrNumbers[natts++] = chunkrel->rd_att->attrs[j]->attnum;
+				ii->ii_KeyAttrNumbers[natts++] = attno;
```

Comment:
should this be ii->ii_KeyAttrNumbers[i] = attno; I don't get what `natts` represents here